pkg/automationconfig/automation_config_builder.go

```diff
@@ -217,31 +217,36 @@ func (b *Builder) setFeatureCompatibilityVersionIfUpgradeIsHappening() error {
 }
 
 func (b *Builder) Build() (AutomationConfig, error) {
-	hostnames := make([]string, b.members+b.arbiters)
+	if err := b.setFeatureCompatibilityVersionIfUpgradeIsHappening(); err != nil {
+		return AutomationConfig{}, fmt.Errorf("can't build the automation config: %s", err)
+	}
+
+	hostnames := make([]string, 0, b.members+b.arbiters)
 
 	// Create hostnames for data-bearing nodes. They start from 0
 	for i := 0; i < b.members; i++ {
-		hostnames[i] = fmt.Sprintf("%s-%d.%s", b.name, i, b.domain)
+		hostnames = append(hostnames, fmt.Sprintf("%s-%d.%s", b.name, i, b.domain))
 	}
 
 	// Create hostnames for arbiters. They are added right after the regular members
-	for i := b.members; i < b.arbiters+b.members; i++ {
+	for i := 0; i < b.arbiters; i++ {
 		// Arbiters will be in b.name-arb-svc service
-		hostnames[i] = fmt.Sprintf("%s-arb-%d.%s", b.name, i-b.members, b.arbiterDomain)
+		hostnames = append(hostnames, fmt.Sprintf("%s-arb-%d.%s", b.name, i-b.members, b.arbiterDomain))
 	}
 
 	members := make([]ReplicaSetMember, b.members+b.arbiters)
 	processes := make([]Process, b.members+b.arbiters)
 
-	if err := b.setFeatureCompatibilityVersionIfUpgradeIsHappening(); err != nil {
-		return AutomationConfig{}, fmt.Errorf("can't build the automation config: %s", err)
-	}
-
 	dataDir := DefaultMongoDBDataDir
 	if b.dataDir != "" {
 		dataDir = b.dataDir
 	}
 
+	fcv := versions.CalculateFeatureCompatibilityVersion(b.mongodbVersion)
+	if b.fcv != "" {
```

Comment:
Probably a better way is to check the `len(b.fcv)` here.

Reply:
Yes, it is. If you can, please help me see if there are any other problems. I will modify it once

Reply:
That's the only spot I found. Otherwise, LGTM